### PR TITLE
Rename IMetadata and MetadataBase to be general purpose Annotatable

### PR DESCRIPTION
**Pull request by mikary**
Changes for #1778 
@rowanmiller @bricelam

### CODE REVIEW

File: src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using JetBrains.Annotations;
+using Microsoft.Data.Entity.Utilities;
+
+namespace Microsoft.Data.Entity.Infrastructure
+{
+    public class AnnotationNotFoundException : Exception
```
Comment (anpete) [inline src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs]: We need a custom exception type for this?

File: src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        public AnnotationNotFoundException()
+        {
+        }
+
+        public AnnotationNotFoundException([NotNull] string message)
+            : base(message)
+        {
+            Check.NotEmpty(message, nameof(message));
```
Comment (anpete) [inline src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs]: Check inline when arg is passed to base.

File: src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        public AnnotationNotFoundException([NotNull] string message, [CanBeNull] Exception innerException)
+            : base(message, innerException)
+        {
+            Check.NotEmpty(message, nameof(message));
```
Comment (anpete) [inline src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs]: Check inline when arg is passed to base.

**Comment by anpete** [comment]
:shipit:

File: src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs (hunk 1 above)
Comment (mikary) [inline src/EntityFramework.Core/Infrastructure/AnnotationNotFoundException.cs]: Not necessarily, the ```ModelItemNotFoundException``` seemed inappropriate for a general purpose component, but an ```InvalidOperationException``` is probably just as good.

**Pull request merged**